src/cascadia/TerminalControl/ThrottledFunc.h

```diff
@@ -0,0 +1,134 @@
+/*++
+Copyright (c) Microsoft Corporation
+Licensed under the MIT license.
+
+Module Name:
+- ThrottledFunc.h
+
+Abstract:
+- This module is a helper to throttle actions.
+- The action is defined by the user of that helper. It could be updating a
+  a file, fetching something from the disk, etc.
+- You give it a minimum delay between two actions and every time you want to
+  start a new action, you call the `Run` method with an argument.
+- Then the function give in the constructor will be called, but only when
+  enough time passed since the last call.
+--*/
+
+#pragma once
+#include "pch.h"
+
+#include "ThreadSafeOptional.h"
+
+template<typename T>
+class ThrottledFunc : public std::enable_shared_from_this<ThrottledFunc<T>>
+{
+public:
+    using Func = std::function<void(T arg)>;
+
+    ThrottledFunc(Func func, winrt::Windows::Foundation::TimeSpan minimumDelay) :
+        _func{ func },
+        _minimumDelay{ minimumDelay }
+    {
+    }
+
+    // Method Description:
+    // - Runs the function later with the specified argument, except if `Run`
+    //   is called again before with a new argument, in which case the new
+    //   argument will be instead.
+    // - For more information, read the Abstract section in the header file.
+    // Arguments:
+    // - arg: the argument
+    // Return Value:
+    // - <none>
+    winrt::fire_and_forget Run(T arg)
+    {
+        if (!_nextArg.Emplace(arg))
+        {
+            // already pending
+            return;
+        }
+
+        auto weakThis = this->weak_from_this();
+
+        const auto waitTime = GetNextWaitTime();
+        if (waitTime.has_value())
+        {
+            co_await winrt::resume_after(waitTime.value());
+        }
+
+        if (auto that{ weakThis.lock() })
+        {
+            // No need to lock because there cannot be another thread that is
+            // reading/writing to the value because `_nextArg` has a value so
+            // other calls to `Run` early return.
+            that->_lastRun = std::chrono::high_resolution_clock::now();
+
+            auto arg = that->_nextArg.Take();
+            if (arg.has_value())
+            {
+                that->_func(arg.value());
+            }
+            else
+            {
+                // should not happen
+            }
+        }
+    }
+
+    template<typename F>
+    void ModifyPending(F f)
+    {
+        _nextArg.ModifyValue(f);
+    }
+
+private:
+    // Method Description:
+    // - Computes the time that the code should wait before doing the next action,
+    //   in order to throttle actions with the specified throttle time. If the code
+    //   should not wait, then an empty value is returned.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - the time to wait before the next action or an empty value for no wait
+    std::optional<winrt::Windows::Foundation::TimeSpan> GetNextWaitTime()
```

Comment:
I'm not entirely sure if it's worth the hassle to implement it this way compared to a simpler "just delay it 8ms every time".
It's a beautiful solution - that's no question - but does it _really_ provide a noticeable improvement?

Reply:
Yes you're right. At first, I added this because I didn't want to regress the first update by 8 ms. But now I think that it makes more sense to always delay the updates by 8 ms in case we have more than 1 update very quickly so that the user only gets to see the last, instead of the first and then the last (to prevent the scrollbar from jumping too much).